Fix restore from backup execution path to use context from caller (#12828)
vitessio/vitess: changelog/17.0/17.0.0/summary.md

    - [VTAdmin web migrated from create-react-app to vite](#migrated-vtadmin)
    - [Keyspace name validation in TopoServer](#keyspace-name-validation)
    - [Shard name validation in TopoServer](#shard-name-validation)

>     - [VtctldClient call to RestoreFromBackup using correct context](#VtctldClient-RestoreFromBackup)

Author: fixed

Shard names may no longer contain the forward slash ("/") character, and TopoServer's `CreateShard` method returns an error if given such a name.

#### <a id="VtctldClient-RestoreFromBackup"> VtctldClient call to RestoreFromBackup using correct context

> The VtctldClient calls for RestoreFromBackup initiate an asynchronous process to restore data from either the latest backup or the closest one before the specified backup-timestamp.

Reviewer: ```suggestion
The VtctldClient command RestoreFromBackup initiates an asynchronous process on the specified tablet to restore data from either the latest backup or the closest one before the specified backup-timestamp.
```

> Prior to v17, this asynchronous process could run indefinitely in the background since it was called using an empty background. In v17 [PR#12830](https://github.com/vitessio/vitess/issues/12830),

Author: done

> this behavior was changed to use the same context with which the client called the RestoreFromBackup command, which uses action_timeout to wait for any command to finish.

Reviewer: ```suggestion
this behavior was changed to use a context with a timeout of `action_timeout`.
```

Author: done

If you are using VtctldClient to initiate a restore, make sure you provide an appropriate value for action_timeout to give enough time for the restore process to complete.

> Otherwise, the restore will throw an error if the context expires before it completes.

Reviewer: What happens to the tablet state? does it go back to its previous tablet_type?

Author: Restore is tricky. It does not go back to original state and it keeps sentinel file which is indication that last restore failed. Since we cancel in the middle of copy , depends upon where we left it give error during the restore while calling ChangeTabletType()->updateLocked->ts.canServe

Author: part of it is due to fact that during restore we first delete all the files. 
	logger.Infof("Restore: deleting existing files")
	if err := removeExistingFiles(cnf); err != nil {
		return err
	}
This prevent us from being coming back to original state.

Author: So I verified this again .. in topo we go back to 'REPLICA' state but in VTAdmin (which calls `show vitess_tablets`) will show `RESTORE`. Reason being as stated above mysql will be down due to incomplete set of files.

Reviewer: If this is really the behavior then it's a healthcheck issue. When the tablet goes back to REPLICA, it should be sending a healthcheck update to vtgate which updates its cache, which is what is used by `show vitess_tablets`. Can you file a separate issue for this? It's so far from this PR's issue that we should not try to fix it in this PR.

Author: for number 2, since restore fail and we stop the MySql , therefore it lost .sock file so we weren't able to connect and do any change to db.. the topo is refresh to replica value (therefore vtctlclient give you 'replica') and as I mentioned show vitess_tablet continue to show restore.

Author: bug filed
https://github.com/vitessio/vitess/issues/12945
